pkg/compose/hash.go

```diff
@@ -40,3 +46,122 @@ func ServiceHash(o types.ServiceConfig) (string, error) {
 	}
 	return digest.SHA256.FromBytes(bytes).Encoded(), nil
 }
+
+// ServiceConfigsHash computes the configuration hash for service configs.
+func ServiceConfigsHash(project *types.Project, serviceConfig types.ServiceConfig) (string, error) {
+	bytes := make([]byte, 0)
+	for _, config := range serviceConfig.Configs {
+		file := project.Configs[config.Source]
+		b, err := createTarForConfig(project, types.FileReferenceConfig(config), types.FileObjectConfig(file))
+
+		if err != nil {
+			return "", err
+		}
+
+		bytes = append(bytes, b.Bytes()...)
+	}
+
+	return digest.SHA256.FromBytes(bytes).Encoded(), nil
+}
+
+// ServiceSecretsHash computes the configuration hash for service secrets.
+func ServiceSecretsHash(project *types.Project, serviceConfig types.ServiceConfig) (string, error) {
+	bytes := make([]byte, 0)
+	for _, config := range serviceConfig.Secrets {
+		file := project.Secrets[config.Source]
+		b, err := createTarForConfig(project, types.FileReferenceConfig(config), types.FileObjectConfig(file))
+
+		if err != nil {
+			return "", err
+		}
+
+		bytes = append(bytes, b.Bytes()...)
+	}
+
+	return digest.SHA256.FromBytes(bytes).Encoded(), nil
+}
+
+func createTarForConfig(
+	project *types.Project,
+	serviceConfig types.FileReferenceConfig,
+	file types.FileObjectConfig,
+) (*bytes.Buffer, error) {
+	// fixed time to ensure the tarball is deterministic
+	modTime := time.Unix(0, 0)
+	content := make([]byte, 0)
+
+	if file.Content != "" {
+		content = []byte(file.Content)
+	} else if file.Environment != "" {
+		env, ok := project.Environment[file.Environment]
+		if !ok {
+			return nil, fmt.Errorf(
+				"environment variable %q required by file %q is not set",
+				file.Environment,
+				file.Name,
+			)
+		}
+		content = []byte(env)
+	} else if file.File != "" {
+		var err error
+		content, err = readPathContent(file.File)
+
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if len(content) == 0 {
+		return nil, fmt.Errorf("config %q is empty", file.Name)
+	}
+
+	if serviceConfig.Target == "" {
+		serviceConfig.Target = "/" + serviceConfig.Source
+	}
+
+	b, err := utils.CreateTar(content, serviceConfig, modTime)
+	if err != nil {
+		return nil, err
+	}
+
+	return &b, nil
+}
+
+func readPathContent(path string) ([]byte, error) {
+	content := make([]byte, 0)
+
+	// Check if the path is a directory
+	info, err := os.Stat(path)
+	if err != nil {
+		return nil, fmt.Errorf("error accessing path %q: %v", path, err)
+	}
+
+	if info.IsDir() {
+		// If it's a directory, read all files and concatenate their contents
+		err = filepath.Walk(path, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if !info.IsDir() {
+				fileContent, err := os.ReadFile(path)
```

Comment:
this will not detect file being rename. Maybe better to create a tar stream from file/dir content

Reply:
@ndeloof it creates a tar from the content here pkg/compose/hash.go:122. 

`b, err := utils.CreateTar(content, serviceConfig, modTime)`

Reply:
@ndeloof is there any reason to create tar inside a tar?   the `createTarForConfig` func calls this `readPathContent` func and creates a tar with the path content

Reply:
here you are creating a tar with all content appended as []byte, this doesn't reflect the filesystem structure for intput. You have to create one tar entry per file with file/folder name.

Reply:
@ndeloof you're right, thank you! Updated the code, now it creates a tar file for a given path and keeps file structure